### PR TITLE
conda recipe and binstar builds

### PR DESCRIPTION
**Pull request by cowlicks**
You can try out the conda package:

```bash
$ conda install -c cowlicks dask
```

Note that I had to use the `build.sh` script to install from bcolz master. This works but it is a dirty hack. It should be removed and bcolz should be added in the `meta.yml`'s `run` field whenever there is a working (with dask) bcolz conda package.

I tried submitting the binstar build but it said no workers were available, so CI is probably not working yet? I have binstar build authorized to test gh/cowlicks/dask. I'm not sure if it is supposed to tests PR's or pushes or if I'm supposed to see a thing like travis-ci has or what. We would need to authorize binstar build to access gh/ContinuumIO/dask to do its thing.

### CODE REVIEW

**Comment by mrocklin** [comment]
I've added you to the blaze user on binstar build.  You should have build queues for all of Linux/Mac/Windows through that user.

File: conda.recipe/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package:
+  name: dask
+  version: 'nightly'
+
+build:
+  number: 1
```
Comment (mrocklin) [inline conda.recipe/meta.yaml]: Version numbering can be tricky.  I strongly recommend stealing most of this from odo

File: conda.recipe/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - toolz
+    - psutil
+    - chest
+    - heapdict
+    - readline
```
Comment (mrocklin) [inline conda.recipe/meta.yaml]: What is readline?

File: conda.recipe/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    - cython  # to compile bcolz
+    - numpy   # bcolz dependency
+
+  run:
+    - python
+    - toolz
+    - numpy
+    - pandas
+    - pyzmq
+    - networkx
```
Comment (mrocklin) [inline conda.recipe/meta.yaml]: I would remove networkx as a dependency.  It's really only for visualization.

File: .binstar.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# Build results are split into two categories: artifacts and targets
+# You may omit either key and stiff have a successful build
+# They may be a string, list and contain any bash glob
+#===============================================================================
+
+## Build Targets: Upload these files to your binstar package
+## build targets may be a list of files (globs allows) to upload
+## The special build targets 'conda' and 'pypi' may be used to 
+## upload conda builds  
+## e.g. conda is an alias for /opt/anaconda/conda-bld/<os-arch>/*.tar.bz2
```
Comment (mrocklin) [inline .binstar.yml]: We might want to scrape out the automatic comments.  Again, recommend stealing from odo

File: conda.recipe/meta.yaml (hunk 2 above)
Comment (cowlicks) [inline conda.recipe/meta.yaml]: I don't know. It was installed in my environment for some reason. But everything tests out fine with it removed.

File: conda.recipe/meta.yaml (hunk 3 above)
Comment (cowlicks) [inline conda.recipe/meta.yaml]: okay, networkx tests need to importorskipped

File: conda.recipe/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    - heapdict
+    - readline
+
+test:
+  requires:
+    - pytest
+
+about:
+  home: https://github.com/ContinuumIO/dask
+  summary: Dask is a task scheduler.
```
Comment (cowlicks) [inline conda.recipe/meta.yaml]: @mrocklin suggestions?

Comment (mrocklin) [inline conda.recipe/meta.yaml]: Task scheduling and blocked algorithms for parallel processing

**Comment by cowlicks** [comment]
Builds here http://alpha.binstar.org/blaze/dask/builds/1

**Comment by cowlicks** [comment]
Some weirdness going on: http://alpha.binstar.org/blaze/dask/builds/1/6#L1108

**Comment by cowlicks** [comment]
Binstar build might need `.binstar.yml` to be in dask master. So we might just have to merge this to test it.

File: .binstar.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  - python -V
+
+install:
+  - conda config --add channels blaze
+
+script:
+  - conda build conda.recipe
+
+env:
+  - LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
```
Comment (mrocklin) [inline .binstar.yml]: Do we need this env variable in dask?

File: .binstar.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+iotimeout: 120
+
+build_targets: conda
+
+notifications:
+  email:
+    recipients: ['mrocklin@continuum.io', 'phillip.cloud@continuum.io',
```
Comment (mrocklin) [inline .binstar.yml]: @cpcloud do you want notifications on dask?

**Comment by mrocklin** [comment]
Other than small questions this looks good to me

File: .binstar.yml (hunk 6 above)
Comment (cowlicks) [inline .binstar.yml]: I don't know. I just copied it from odo. I'm not sure what it does.

File: .binstar.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                 'bgriffith@continuum.io']
+
+---
+platform: win-32
+engine: python=2.6
+env: LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
+exclude: true
```
Comment (cowlicks) [inline .binstar.yml]: I'm not sure if we need these either.

File: .binstar.yml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  - python=2.6
+  - python=2.7
+  - python=3.3
+  - python=3.4
+
+before_script:
+  - python -V
+
+install:
+  - conda config --add channels blaze
```
Comment (mrocklin) [inline .binstar.yml]: we probably also don't need this

File: .binstar.yml (hunk 6 above)
Comment (mrocklin) [inline .binstar.yml]: Probably necessary for handling strange formats in some database.  Suggest removal

File: conda.recipe/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  git_url: ../
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - toolz
+    - psutil
+    - chest
+    - heapdict
```
Comment (mrocklin) [inline conda.recipe/meta.yaml]: heapdict is a dependency of chest.  We should be able to depend on it implicitly

**Comment by cowlicks** [comment]
I actually did get the building to work but  there are a few problems. Builds here https://binstar.org/cowlicks/dask/builds

**Comment by mrocklin** [comment]
I wonder if @cpcloud  has suggestions.  That error reminds of errors I had when I was versioning incorrectly.

**Comment by cowlicks** [comment]
@mrocklin some of the binstar folks are helping me out

File: .binstar.yml (hunk 7 above)
Comment (cpcloud) [inline .binstar.yml]: Not for binstar build errors

File: .binstar.yml (hunk 8 above)
Comment (cpcloud) [inline .binstar.yml]: This was to deal with a default encoding that wasn't specified on binstar on osx and linux64 IIRC. I'd leave it off for now and add it only if necessary.

**Comment by cowlicks** [comment]
Tests passing https://binstar.org/blaze/dask/builds/14/4 still some work to do though

**Comment by mrocklin** [comment]
In https://binstar.org/blaze/dask/settings/ci I've updated the build queue to `build-binstar-continuum`.  This should now start testing against other platforms.

**Comment by mrocklin** [comment]
Looks like we've exposed bugs in the distributed scheduler on Mac and Linux-32

File: conda.recipe/build.sh [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+pip install git+https://github.com/blosc/bcolz --upgrade
+$PYTHON setup.py install
```
Comment (cowlicks) [inline conda.recipe/build.sh]: How would this, translate to a windows build script?

File: conda.recipe/run_test.sh [hunk 12]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+# is $SRC_DIR right?
+py.test -vv $SRC_DIR
```
Comment (cowlicks) [inline conda.recipe/run_test.sh]: I'm also not sure how to rewrite this as a `.bat` script.

**Comment by cowlicks** [comment]
I'm confused/surprised the windows boxes build from a bash script. They are not running tests, I think I need a `run_test.bat` script for that. I think it should be a simple translation but I'm not even sure where to begin.

File: conda.recipe/run_test.sh (hunk 12 above)
Comment (mrocklin) [inline conda.recipe/run_test.sh]: In `odo` we include this as part of the conda.recipe/meta.yaml file

```
test:
      commands:
        - py.test --pyargs odo
```

**Comment by cowlicks** [comment]
https://binstar.org/blaze/dask/builds/18/12#L1809

Windows tests are running, but the all fail with import errors

**Comment by cowlicks** [comment]
Okay windows tests are passing but there is a windows bug causing a test to hang.

https://binstar.org/blaze/dask/builds/19/16#L2986

The test is probably test_bag.py::test_map_function_with_multiple_arguments

https://binstar.org/blaze/dask/builds/19/4#L2223

**Comment by cowlicks** [comment]
There is an osx bug for "Too many open files" https://binstar.org/blaze/dask/builds/19/8#L1592

**Comment by cowlicks** [comment]
It looks like binstar build is configured properly. Things are getting built and tested. So I think this is ready pending review and removal of some debug code.

**Comment by mrocklin** [comment]
> There is an osx bug for "Too many open files" https://binstar.org/blaze/dask/builds/19/8#L1592

That's not entirely surprising.  I can resolve that.

File: conda.recipe/meta.yaml (hunk 10 above)
Comment (mrocklin) [inline conda.recipe/meta.yaml]: Can now remove both

File: conda.recipe/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,49 @@
+package:
+  name: dask
+  # GIT_DESCRIBE_TAG is supposed to return the latest version (0.4.0) but it is
+  # not working for some reason. And PKG_VERSION is incorrect (it says 1.0.1k)
+  version: {{ environ.get('GIT_DESCRIBE_TAG', '0.4.0')}}
```
Comment (mrocklin) [inline conda.recipe/meta.yaml]: Do we know why this is?  Does it work locally and not on binstar?  Perhaps our tags are mismanaged somehow?  If we can't figure this out we should bug the conda folk.

File: conda.recipe/meta.yaml (hunk 10 above)
Comment (mrocklin) [inline conda.recipe/meta.yaml]: Oh, actually if we're adding in all possible dependencies then we might as well keep it in.  Probably should add pyzmq as well.

**Comment by mrocklin** [comment]
> There is an osx bug for "Too many open files" https://binstar.org/blaze/dask/builds/19/8#L1592

I hope that this is now resolved.  I don't have an osx machine lying around to test it though.  Relying on this and binstar-build to guide me here.

File: conda.recipe/meta.yaml (hunk 13 above)
Comment (cowlicks) [inline conda.recipe/meta.yaml]: I think this only works when the build is triggered by a git hook. Not when I trigger it manually. Right @malev ?

Comment (mrocklin) [inline conda.recipe/meta.yaml]: Oh, is it possible that you don't have jinja2 (or whatever templating system this is) installed locally?

**Comment by mrocklin** [comment]
I suggest removing comments, cleaning up a bit, and then merging.  We can iterate on this a bit once binstar-build has easier access to it.

File: conda.recipe/meta.yaml (hunk 10 above)
Comment (cowlicks) [inline conda.recipe/meta.yaml]: Is pyzmq a "build" dependency too? I have it under the `run` field.

I think that the build dependencies are removed after the package is built. Then the runtime dependencies are installed.

File: conda.recipe/meta.yaml (hunk 13 above)
Comment (cowlicks) [inline conda.recipe/meta.yaml]: Now this appears to be working. I'm not sure what changed.

**Comment by mrocklin** [comment]
This looks good to me.  OK to merge?

**Comment by mrocklin** [comment]
In builds like [this one](https://binstar.org/blaze/dask/builds/20/9) we have test failures but binstar-build gives us a green light.

**Comment by mrocklin** [comment]
Versions are a bit wonky still.  Looking at https://binstar.org/blaze/dask/builds/20/4


Build target is linux-64/dask-None-py26_0.tar.bz2

2276# $ binstar upload /home/binstar/miniconda/conda-bld/linux-64/dask-None-py26_0.tar.bz2

**Comment by mrocklin** [comment]
Thoughts on what's going on with windows 64 builds?

https://binstar.org/blaze/dask/builds/23/16

**Comment by cowlicks** [comment]
@mrocklin I'm not sure. It doesn't even look like dask is being built yet. Maybe a problem with versioning and naming the package?

**Pull request merged**